RandLAPACK/drivers/rl_cqrrpt_gpu.hh

```diff
@@ -0,0 +1,395 @@
+        qrcp_t_stop = high_resolution_clock::now();
+        rank_reveal_t_start = high_resolution_clock::now();
+    }
+
+    /// Using naive rank estimation to ensure that R used for preconditioning is invertible.
+    /// The actual rank estimate k will be computed a posteriori. 
+    /// Using R[i,i] to approximate the i-th singular value of A_hat. 
+    /// Truncate at the largest i where R[i,i] / R[0,0] >= eps.
+    for(i = 0; i < n; ++i) {
+        if(std::abs(A_hat[i * d + i]) / std::abs(A_hat[0]) < eps_initial_rank_estimation) {
+            k = i;
+            break;
+        }
+    }
+    this->rank = k;
+
+    if(this -> timing)
+        rank_reveal_t_stop = high_resolution_clock::now();
+
+    // Allocating space for a preconditioner buffer.
+    T* R_sp  = ( T * ) calloc( k * k, sizeof( T ) );
+    /// Extracting a k by k upper-triangular R.
+    lapack::lacpy(MatrixType::Upper, k, k, A_hat, d, R_sp, k);
+    /// Extracting a k by n R representation (k by k upper-triangular, rest - general)
+    lapack::lacpy(MatrixType::Upper, k, k, A_hat, d, R, ldr);
+    lapack::lacpy(MatrixType::General, k, n - k, &A_hat[d * k], d, &R[n * k], ldr);
+
+    if(this -> timing)
+        a_mod_piv_t_start = high_resolution_clock::now();
+
+    // Swap k columns of A with pivots from J
+    blas::copy(n, J, 1, J_buf.data(), 1);
+    util::col_swap(m, n, k, A, lda, J_buf);
+
+    if(this -> timing) {
+        a_mod_piv_t_stop = high_resolution_clock::now();
+        a_mod_trsm_t_start = high_resolution_clock::now();
+    }
+
+    /******************************GPU REGION BEGIN*********************************/
+    // The reasons for using GPUs for this part only ar the following: 
+    // 1. There is no geqp3 available in any GPU linalg libraries. 
+    //    We could port HQRRP to GPUs, but that takes additional time.
+    // 2. There are no lacpy functions on any GPU linalg libraries. Those, however, 
+    //    can be substituted with blas::copy.
+    // 3. We do not have GPU-based skething support in RandBLAS at the moment.
+    //
+    // If the above points would be resolved, we may perform the entirety of CQRRPT on the device.
+    //
+    // Allocating device data & performing copies:
+    T* A_device;
+    T* R_device;
+    T* R_sp_device;
+
+    // Variables for a posteriori rank estimation.
+    int64_t new_rank;
+    T running_max, running_min, curr_entry;
+
+    using lapack::device_info_int;
+    blas::Queue blas_queue(0);
+    lapack::Queue lapack_queue(0);
+    device_info_int* d_info = blas::device_malloc< device_info_int >( 1, lapack_queue );
+    
+    cudaMalloc(&A_device, m * n * sizeof(T));
+    cudaMalloc(&R_device, ldr * n * sizeof(T));
+    cudaMalloc(&R_sp_device, k * k * sizeof(T));
+
+    cudaMemcpy(A_device, A, m * n * sizeof(T), cudaMemcpyHostToDevice);
+    cudaMemcpy(R_device, R, ldr * n * sizeof(T), cudaMemcpyHostToDevice);
+    cudaMemcpy(R_sp_device, R_sp, k * k * sizeof(T), cudaMemcpyHostToDevice);
```

Comment:
Please open a GitHub issue after merging this PR. Have the issue mention a TODO of trying to perform some of the data-transfer or copying asynchronously while sketching is happening. It might end up being faster to perform the pivoting on the GPU rather than wait until this point to start the data transfer.

Reply:
Sure
To be fair though, the CQRRPT GPU code is pretty bad overall, and was only serving a purpose of me learning how to use basic GPU stuff.

Reply:
Added an issue.